### ISSUE TITLE
有没有使用其他框架推理的计划？比如onnxruntime和openvino以及trt

### ISSUE THREAD

**Issue by xinsuinizhuan** [UNeedCryDear/yolov5-seg-opencv-onnxruntime-cpp#5]
我使用onnxruntime推理出来的结果，很烂
这是使用您的dnn推理出来的：
![图片](https://user-images.githubusercontent.com/40679769/205420834-1357077b-b52a-4867-91a6-7bc2ccf13a3b.png)

这是我自己写的使用onnxruntime推理出来的：
![图片](https://user-images.githubusercontent.com/40679769/205420887-28f503a9-b7fe-4b7a-bcca-5b66db30db0a.png)

最多的区别是预处理部分和推理完后数据格式部分：
onnxruntime预处理部分：
       1、转rgb格式：
	cv::Mat canvas;
	cv::cvtColor(mat_rs, canvas, cv::COLOR_BGR2RGB);
       2、然后归一化
       normalize_inplace(canvas, 0, 1/255);
void normalize_inplace(cv::Mat& mat_inplace, const float* mean, const float* scale)
{
	if (mat_inplace.type() != CV_32FC3) mat_inplace.convertTo(mat_inplace, CV_32FC3);
	for (unsigned int i = 0; i < mat_inplace.rows; ++i)
	{
		cv::Vec3f* p = mat_inplace.ptr<cv::Vec3f>(i);
		for (unsigned int j = 0; j < mat_inplace.cols; ++j)
		{
			p[j][0] = (p[j][0] - mean[0]) * scale[0];
			p[j][1] = (p[j][1] - mean[1]) * scale[1];
			p[j][2] = (p[j][2] - mean[2]) * scale[2];
		}
	}
}

       3、推理部分结果处理：
       由于推理出来的数据是float*格式的，但是dnn推理出来的结果直接是cv::Mat格式
       对于淹码数据部分：
       const float* mask_buffer = outputTensors[1].GetTensorMutableData<float>();
       然后，把它格式化成4维数据：
	int mask_size[4] = { mask_bat, mask_channel, mask_rows, mask_cols };
	cv::Mat mask_protos = cv::Mat::zeros(4, mask_size, CV_32FC1);
	memcpy(mask_protos.data, mask_buffer, mask_bat * mask_channel * mask_rows * mask_cols);

       其他部分，跟dnn一样，但是结果就是不一样，很奇怪

**Comment by UNeedCryDear** [comment]
>  由于推理出来的数据是float*格式的，但是dnn推理出来的结果直接是cv::Mat格式

格式什么的不影响，你可以看我也都是直接转为指针处理的。
> 2、然后归一化

我一般不会像你这样单独处理，因为我测试过了，这种方案速度不行，一般用的opencv的blobFromImage()这个API，将数据转成4维并且归一化，之后这个的输出就是blob，shape:[1,3,640,640],然后将他的指针通过ort的CreateTensor接口的第二个参数（这里刚好要求指针）赋值给ort的输入即可，大概类似下面这种做法。
```
		Mat blob = blobFromImages(img, 1 / 255.0, inputSize, Scalar(0, 0, 0), true, false);
		auto input=Ort::Value::CreateTensor<float>(memoryInfo, (float*)blob.data, inputTensorSize, inputTensorShape.data(),inputTensorShape.size())
```

**Comment by xinsuinizhuan** [comment]
> memcpy改下？这个是字节复制的，float类型是4个字节，你这里的长度都不对，结果肯定有问题啊

应该怎么改？改成什么操作？一个一个复制？

**Comment by xinsuinizhuan** [comment]
> memcpy改下？这个是字节复制的，float类型是4个字节，你这里的长度都不对，结果肯定有问题啊

感谢您。已经解决了！

**Comment by UNeedCryDear** [comment]
已经新增了onnxruntime部署，至于trt，已经有人写了我就不继续了。

**Comment by ghmchen** [comment]
> 已经新增了onnxruntime部署，至于trt，已经有人写了我就不继续了。

Hello UNeedCryDear, I would like to thank you for your great contributions in C++ deployment of yolov5 and v7.  Your work have benfited many people like me. Do you have a plan to make c++ instance segmentation (yolov7 u7 branch) work? Thanks.

**Comment by UNeedCryDear** [comment]
> Hello UNeedCryDear, I would like to thank you for your great contributions in C++ deployment of yolov5 and v7. Your work have benfited many people like me. Do you have a plan to make c++ instance segmentation (yolov7 u7 branch) work? Thanks.

yolov7我之前试过mask分支的部署，由于部分算子和部分网络结构opencv无法支持，并且windows下面的detectron2不好弄，后面我重装系统之后我就没继续做了。至于你说yolov7-u7分支，我没试过，不过看他的结构和readme是基于yolov5的，你可以试试看下yolov5-seg的onnxruntime版本能不能运行起来导出的onnx模型。至于opencv的版本，我估计希望不大，局限性太大了，不如onnxruntime

**Comment by ghmchen** [comment]
根据你 github提供的代码，yolov5-seg 可以导出onnx 模型，并在opencvDNN C++ 上运行。 yolov7-u7 分支 我可以导出 onnx 模型，在python下可以成功运行（predict.py).  我需要设法把yolov7-u7 分支 导出的 onnx 模型。运行在C++ openCV DNN 上。 谢谢回复。

**Comment by UNeedCryDear** [comment]
> 根据你github提供的代码，yolov5-seg可以导出onnx模型，并在opencvDNN C++上运行。yolov7-u7分支持我可以导出onnx模型，在python下可以成功运行（predict.py）。我需要设置方法把yolov7-u7分支持导出的onnx模型。运行在C++ openCV DNN上。谢谢回复。

我看了一下yolov7-u7分支，这不就是yolov5的v6.2版本吗？那不就没啥差别，直接导出就可以跑onnx了啊，你试下？

**Comment by ghmchen** [comment]
我按照导出yolov5 的方法，成功导出 yolov7-seg u7 的onnx.
C:\yolo\yolov7\seg\export.py 
--weights
C:/yolo/yolov7/seg/runs/train-seg/torborweb_model/weights/best.pt
--img
640
--batch
1
--opset
12
--include
onnx

用predict.py 测试 导出的onnx，可以成功输出结果。
C:\yolo\yolov7\seg\segment\predict.py
--img
640
--weights
C:\yolo\yolov7\seg\runs\train-seg\torborweb_model\weights\best.onnx
--source
"D:\temp\probe2021081806107\img_baddata_256\L0_210_5.png"
![image](https://user-images.githubusercontent.com/46462888/208814330-dd1a7792-b827-4de5-849b-e6bfec197bfc.png)
但是当用你的yolov5-seg-opencv-onnxruntime-cpp 运行 此onnx 时，出现以下错误：
[ INFO:0] global D:\opencv453\opencv-4.5.3\modules\dnn\src\onnx\onnx_importer.cpp (429) cv::dnn::dnn4_v20210608::ONNXImporter::populateNet DNN/ONNX: loading ONNX v7 model produced by 'pytorch':1.12.1. Number of nodes = 554, inputs = 1, outputs = 5
OpenCV(4.5.3) Error: Assertion failed (inputs.size()) in cv::dnn::dnn4_v20210608::Layer::getMemoryShapes, file D:\opencv453\opencv-4.5.3\modules\dnn\src\dnn.cpp, line 5298
[ERROR:0] global D:\opencv453\opencv-4.5.3\modules\dnn\src\onnx\onnx_importer.cpp (2127) cv::dnn::dnn4_v20210608::ONNXImporter::handleNode DNN/ONNX: ERROR during processing node with 1 inputs and 1 outputs: [Identity]:(onnx::Reshape_910)
[ INFO:0] global D:\opencv453\opencv-4.5.3\modules\dnn\src\onnx\onnx_importer.cpp (2131) cv::dnn::dnn4_v20210608::ONNXImporter::handleNode     Input[0] = 'onnx::Reshape_870'
[ INFO:0] global D:\opencv453\opencv-4.5.3\modules\dnn\src\onnx\onnx_importer.cpp (2135) cv::dnn::dnn4_v20210608::ONNXImporter::handleNode     Output[0] = 'onnx::Reshape_910'
OpenCV(4.5.3) Error: Unspecified error (> Node [Identity]:(onnx::Reshape_910) parse error: OpenCV(4.5.3) D:\opencv453\opencv-4.5.3\modules\dnn\src\dnn.cpp:5298: error: (-215:Assertion failed) inputs.size() in function 'cv::dnn::dnn4_v20210608::Layer::getMemoryShapes'
> ) in cv::dnn::dnn4_v20210608::ONNXImporter::handleNode, file D:\opencv453\opencv-4.5.3\modules\dnn\src\onnx\onnx_importer.cpp, line 2146

**Comment by UNeedCryDear** [comment]
torch1.12的话你要改下do_constant_folding=False看下
